packages/backend/src/api/controllers/TransactionController.ts

```diff
@@ -0,0 +1,254 @@
+        //     | 'SENT'
+        //     | 'MINED'
+        //     | 'REVERTED'
+        //     | 'INCLUDED'
+        // }[]
+        stateUpdateId: userTransaction.included?.stateUpdateId,
+      })
+    }
+
+    if (
+      userTransaction.data.type === 'Withdraw' ||
+      userTransaction.data.type === 'WithdrawWithTokenId' ||
+      userTransaction.data.type === 'MintWithdraw'
+    ) {
+      const history = buildRegularHistory(sentTransaction, userTransaction)
+      const data = userTransaction.data
+      const assetHash =
+        data.type === 'Withdraw'
+          ? data.assetType
+          : data.type === 'WithdrawWithTokenId'
+          ? data.assetId
+          : data.assetId
+
+      let recipientEthAddress =
+        data.type === 'Withdraw'
+          ? data.recipient
+          : data.type === 'WithdrawWithTokenId'
+          ? data.recipient
+          : undefined
+
+      if (recipientEthAddress === undefined) {
+        const recipient =
+          await this.userRegistrationEventRepository.findByStarkKey(
+            userTransaction.data.starkKey
+          )
+        // TODO handle lack of recipient address
+        recipientEthAddress = recipient?.ethAddress
+      }
+
+      content = renderRegularWithdrawalPage({
+        user,
+        transactionHash: txHash,
+        recipient: {
+          // TODO don't display starkKey if unknown
+          starkKey: StarkKey.ZERO,
+          // TODO don't display ethereum address if unknown
+          ethereumAddress: recipientEthAddress ?? EthereumAddress.ZERO,
+        },
+        asset: { hashOrId: assetHash },
+        amount: userTransaction.data.quantizedAmount,
+        history,
+        stateUpdateId: userTransaction.included?.stateUpdateId,
+      })
+    }
+
+    if (!content) {
+      return {
+        type: 'not found',
+        content: "Transaction can't be displayed right now...",
+      }
+    }
+    return { type: 'success', content }
+  }
+}
+
+function buildRegularHistory(
+  sentTransaction: SentTransactionRecord | undefined,
+  userTransaction: UserTransactionRecord
+): { timestamp: Timestamp; status: 'SENT' | 'REVERTED' | 'MINED' }[] {
+  const history: {
+    timestamp: Timestamp
+    status: 'SENT' | 'REVERTED' | 'MINED'
+  }[] = []
+  history.push({
+    status: 'MINED',
+    timestamp: userTransaction.timestamp,
+  })
+  if (sentTransaction?.mined?.reverted) {
+    history.push({
+      timestamp: sentTransaction.mined.timestamp,
+      status: 'REVERTED',
+    })
+  }
+  if (sentTransaction) {
+    history.push({
+      timestamp: sentTransaction.sentTimestamp,
+      status: 'SENT',
+    })
+  }
+  return history
+}
+
+function buildForcedActionHistory(
+  sentTransaction: SentTransactionRecord | undefined,
+  userTransaction: UserTransactionRecord
+): {
+  timestamp: Timestamp
+  status: 'SENT' | 'REVERTED' | 'MINED' | 'INCLUDED'
+}[] {
+  const history: {
+    timestamp: Timestamp
+    status: 'SENT' | 'REVERTED' | 'MINED' | 'INCLUDED'
+  }[] = []
```

Comment:
We could extract it to a type. It could extend the regular history type if we decided to extract that as well